### ISSUE TITLE
p2p: DNS and retry handling

### ISSUE THREAD

**Issue by ebuchman** [tendermint/tendermint#1521]
Proposal for simplest path forward to reasonable DNS resolution/retry handling. Closes https://github.com/tendermint/tendermint/issues/869

## NetAddress

Currently just grabs the first result in a lookup.

It should instead randomly select a result from the DNS lookup.

## Persistent Peers

Currently does a couple minutes of attempts, then exponential backoff for a day, using the IP address self reported by the peer in its NodeInfo (ie. not even necessarily the IP we dialed, which may have come from a DNS lookup if the persistent_peer was specified with DNS).

It should be:

- keep a map[ID]string in the Switch mapping persistent peer IDs to their address as `HOST:PORT`, where HOST may be a DNS name or an IP address.
- reconnectForPeer retry/backoff logic stays the same, but uses the DNS every time (ie. via NewNetAddress)
- add the new address to the addrbook before dialing
- note the original address is still in the address book. note we can have a few entries per peerID in the addr book

## Address book

Entries in the address book do not (and should not) contain DNS names.

The same ID can be added to the address book multiple times with the same or different IPs.
Once we MarkGood on the ID, we remove the ID from all new buckets and move it to only one "old" bucket with the address that got marked good. Once marked good, an ID can only be added to new buckets again if it comes with a different IP.

**Comment by ebuchman** [comment]
Note we need to support DNS in ExternalAddress too ala https://github.com/tendermint/tendermint/issues/1845

**Comment by coriolinus** [comment]
My company is working on the operational details of getting a network going. It turns out to be important to us to be able to use DNS to identify where our nodes live, because the nature of cloud servers is that the machine on which you're running can change unpredictably. Using DNS lets us keep the network talking to itself even if the physical machinery on which the nodes run, and therefore the IP address, changes regularly. 

If that is already possible, that's awesome! In that case, I'd appreciate some instructions on how to configure things to use IPs. However, the sense I get is that it is currently impossible, because the address book does not retain any knowledge of a URL, just the IP address.

The issue description here states that this is intentional behavior. On the other hand, more recent comments in https://github.com/tendermint/tendermint/issues/869 (which was closed in favor of this issue) seem to advocate updating the address book to retain the URL, and have the book attempt to re-resolve the URL if the IP-based connection stops working. 

The tag says "help wanted". I'm here to help. I'd really appreciate mentoring instructions, though: whether [`addrbook.go`](https://github.com/tendermint/tendermint/blob/master/p2p/pex/addrbook.go) is the right place to start, or if not, where would be. How to write a unit test exercising the PEX system. What systems depend on the old behavior and will need to be updated appropriately. Etc.

**Comment by xla** [comment]
@coriolinus Thanks for providing context and offering to help. We should re-evaluate the solutiosn proposed up top as any change in the p2p package is quite involved, see #2067 for a lot more information around the topic.

If you come this far and are still determine to move forward, a good first step to understand the current package by reading through the documents of the [spec](https://github.com/tendermint/tendermint/tree/master/docs/spec/p2p). After that I'd do an inventory of all the places DNS lookups are performed and how it affects the dialing and retry. If you gained confidence that you have an idea of how to improve the situation, write up an ADR ([examples](https://github.com/tendermint/tendermint/tree/master/docs/architecture)) with your proposed solution. This will spark a meaningful discussion and hopefully lead to a peer-reviewed plan of what to change where.

Hope this is of some help, let us know if you need more info.

**Comment by nyetwurk** [comment]
Depending on p2p is a non-starter; proper DNS support should be baked in from the start, and not an afterthought. Given the state of IPv4 exhaustion, depending on it being static is a terrible design decision

**Comment by nyetwurk** [comment]
This is unacceptable. The fundamental nature of IP addresses is that they can change, and the DNS is the canonical way to resolve them and address cases where the IP addresses change. This should have been known from the start *during the initial design stages* of tendermint. That it wasn't is a huge oversight and is extremely dismaying.

**Comment by albttx** [comment]
Hello, a quick come back on that since i'm working on a large infrastructure of RPC nodes with [traefik](https://traefik.io) as a TCP loadbalancer.

My IP / Port for the p2p endpoint is fully dynamic on my servers, But i have a traefik upfront that could redirect the connection to the right service.

Traefik listen all TCP connection on port `26656` but then will redirect depending on the DNS to the correct server /IP/port. 
DNS is like `juno-p2p.example.com`, `osmosis-p2p.example.com` ...

> Entries in the address book do not (and should not) contain DNS names.

Why that ?

If the persistent_peers arrive in the address book, the resolutions will fail since it will lose the correct destination.

FYI, i currently not manage to have traefik work even with just 1 node, here is the error
```
6:24PM ERR Error dialing peer err="auth failure: secret conn failed: proto: BytesValue: wiretype end group for non-group" module=p2p
```
Traefik require tls for host recognition, could that be linked ? (if you think, i could create a separate issue)

**Comment by nyetwurk** [comment]
Please do not close this.
